### PR TITLE
fix: disable basic auth for HTTP API

### PR DESCRIPTION
**Pull request by lafirest**
From: [EMQX-8376](https://emqx.atlassian.net/browse/EMQX-8376)
Goal: ensure the API only can be called with the API key or use the bearer token

### CODE REVIEW

File: apps/emqx_dashboard/src/emqx_dashboard.erl [hunk 1]
```diff
@@ -45,6 +45,12 @@
 
 -define(EMQX_MIDDLE, emqx_dashboard_middleware).
 
+-ifdef(TEST).
```
Comment (id) [inline apps/emqx_dashboard/src/emqx_dashboard.erl:48]: I think the right approach is to update tests

File: apps/emqx_dashboard/src/emqx_dashboard.erl [hunk 2]
```diff
@@ -65,8 +71,21 @@ start_listeners(Listeners) ->
         components => #{
             schemas => #{},
             'securitySchemes' => #{
-                'basicAuth' => #{type => http, scheme => basic},
-                'bearerAuth' => #{type => http, scheme => bearer}
+                'basicAuth' => #{
```
Comment (id) [inline apps/emqx_dashboard/src/emqx_dashboard.erl:74]: this can be removed if we fix tests

File: apps/emqx_dashboard/src/emqx_dashboard.erl [hunk 3]
```diff
@@ -35,7 +35,7 @@
 ]).
 
 %% Authorization
--export([authorize/1]).
+-export([authorize/1, basic_authorize/3, api_key_authorize/3]).
```
Comment (id) [inline apps/emqx_dashboard/src/emqx_dashboard.erl:38]: why do we need to export these?

**Comment by terry-xiaoyu** [comment]
IMO, we should **never** remove the basic auth. 

We have been supporting the basic auth from the first 1.0 version of emqx, if there's any concern about security, add warnings to the doc and notice the user not to use it.

File: .github/workflows/run_fvt_tests.yaml [hunk 4]
```diff
@@ -201,12 +201,16 @@ jobs:
           echo "waiting emqx started";
           sleep 10;
         done
+    - name: Get Token
+      timeout-minutes: 10
+      run: |
+        echo $"TOKEN=$(curl --silent -X 'POST' 'http://127.0.0.1:18083/api/v5/login' -H 'accept: application/json' -H 'Content-Type: application/json' -d '{"username": "admin","password": "public"}' | jq -r ".token")" >> $GITHUB_ENV
```
Comment (id) [inline .github/workflows/run_fvt_tests.yaml:207]:
```suggestion
        echo "TOKEN=$(curl --silent -X 'POST' 'http://127.0.0.1:18083/api/v5/login' -H 'accept: application/json' -H 'Content-Type: application/json' -d '{"username": "admin","password": "public"}' | jq -r ".token")" >> $GITHUB_ENV
```

File: changes/v5.0.15/feat-9586.en.md [hunk 5]
```diff
@@ -0,0 +1 @@
+The `basic` authentication which based on the `username:password` has no longer supported anymore, now the API must call with the API key.
```
Comment (id) [inline changes/v5.0.15/feat-9586.en.md:1]:
```suggestion
Basic HTTP Authentication (`username:password`) is not supported by HTTP API anymore, all API calls must use `Bearer` scheme and an API key.
```

Comment (zmstone) [inline changes/v5.0.15/feat-9586.en.md:1]:
```suggestion
Basic auth is no longer allowed for API calls, must use API key instead.
```

**Review by zmstone — CHANGES_REQUESTED** [review]

File: .github/workflows/run_fvt_tests.yaml [hunk 6]
```diff
@@ -201,12 +201,25 @@ jobs:
           echo "waiting emqx started";
           sleep 10;
         done
-    - name: Check cluster
+    - name: Get Token
       timeout-minutes: 10
       run: |
         kubectl port-forward service/${{ matrix.profile }} 18083:18083 > /dev/null &
+
+        while
```
Comment (zmstone) [inline .github/workflows/run_fvt_tests.yaml:209]: add a limit, e.g. 60 seconds.

File: apps/emqx/include/http_api.hrl [hunk 7]
```diff
@@ -16,8 +16,8 @@
 
 %% HTTP API Auth
 -define(WRONG_USERNAME_OR_PWD, 'WRONG_USERNAME_OR_PWD').
--define(WRONG_USERNAME_OR_PWD_OR_API_KEY_OR_API_SECRET,
-    'WRONG_USERNAME_OR_PWD_OR_API_KEY_OR_API_SECRET'
+-define(WRONG_API_KEY_OR_API_SECRET,
```
Comment (zmstone) [inline apps/emqx/include/http_api.hrl:19]:
```suggestion
-define(BAD_API_KEY_OR_SECRET,
```

File: apps/emqx/include/http_api.hrl [hunk 8]
```diff
@@ -16,8 +16,8 @@
...
+    'WRONG_API_KEY_OR_API_SECRET'
```
Comment (zmstone) [inline apps/emqx/include/http_api.hrl:20]:
```suggestion
    'BAD_API_KEY_OR_SECRET'
```

File: apps/emqx/include/http_api.hrl [hunk 9]
```diff
@@ -58,7 +58,7 @@
 %% All codes
 -define(ERROR_CODES, [
     {'WRONG_USERNAME_OR_PWD', <<"Wrong username or pwd">>},
-    {'WRONG_USERNAME_OR_PWD_OR_API_KEY_OR_API_SECRET', <<"Wrong username & pwd or key & secret">>},
+    {'WRONG_API_KEY_OR_API_SECRET', <<"Wrong api key or secret">>},
```
Comment (zmstone) [inline apps/emqx/include/http_api.hrl:61]: use the macro

File: apps/emqx_management/test/emqx_mgmt_api_test_util.erl [hunk 10]
```diff
@@ -109,10 +139,7 @@ do_request_api(Method, Request, Opts) ->
     end.
 
 auth_header_() ->
```
Comment (zmstone) [inline apps/emqx_management/test/emqx_mgmt_api_test_util.erl:141]: nit:
can we try to rename this API?
even auth_header2 is better than auth_header_

Comment (lafirest) [reply apps/emqx_management/test/emqx_mgmt_api_test_util.erl:141]: There are three duplicated HTTP test helper files, and there are many same functions in these files, I plan to unite them into one to reduce the chaos, so this name can be renamed in that PR

File: .github/workflows/run_fvt_tests.yaml (hunk 6 above)
Comment (lafirest) [reply .github/workflows/run_fvt_tests.yaml:209]: I have cutted down the limit from 10 minutes to 1.

File: apps/emqx/include/http_api.hrl [hunk 11]
```diff
@@ -58,7 +56,7 @@
 %% All codes
 -define(ERROR_CODES, [
     {'WRONG_USERNAME_OR_PWD', <<"Wrong username or pwd">>},
-    {'WRONG_USERNAME_OR_PWD_OR_API_KEY_OR_API_SECRET', <<"Wrong username & pwd or key & secret">>},
+    {?BAD_API_KEY_OR_SECRET, <<"Wrong api key or secret">>},
```
Comment (zmstone) [inline apps/emqx/include/http_api.hrl:59]:
```suggestion
    {?BAD_API_KEY_OR_SECRET, <<"Bad api key or secret">>},
```

File: apps/emqx/include/http_api.hrl [hunk 12]
```diff
@@ -58,7 +56,7 @@
 %% All codes
 -define(ERROR_CODES, [
     {'WRONG_USERNAME_OR_PWD', <<"Wrong username or pwd">>},
```
Comment (zmstone) [inline apps/emqx/include/http_api.hrl:58]:
```suggestion
    {'WRONG_USERNAME_OR_PWD', <<"Bad username or password">>},
```

File: apps/emqx/include/http_api.hrl [hunk 13]
```diff
@@ -16,9 +16,7 @@
 
 %% HTTP API Auth
 -define(WRONG_USERNAME_OR_PWD, 'WRONG_USERNAME_OR_PWD').
```
Comment (zmstone) [inline apps/emqx/include/http_api.hrl:18]:
```suggestion
-define(BAD_USERNAME_OR_PWD, 'BAD_USERNAME_OR_PWD').
```

File: apps/emqx/test/emqx_common_test_http.erl [hunk 14]
```diff
@@ -29,6 +29,9 @@
     auth_header/2
 ]).
 
+-define(DEFAULT_API_NAME, <<"default_api_name">>).
```
Comment (zmstone) [inline apps/emqx/test/emqx_common_test_http.erl:32]:
```suggestion
-define(DEFAULT_APP_ID, <<"default_appid">>).
```

File: apps/emqx/test/emqx_common_test_http.erl [hunk 15]
```diff
@@ -29,6 +29,9 @@
...
+-define(DEFAULT_API_SECRET, <<"default_api_secret">>).
```
Comment (zmstone) [inline apps/emqx/test/emqx_common_test_http.erl:33]:
```suggestion
-define(DEFAULT_APP_SECRET, <<"default_app_secret">>).
```

**Review by zmstone — APPROVED** [review]

**Pull request merged**